tracing-opentelemetry/src/main/java/io/micronaut/tracing/opentelemetry/http/OpenTelemetryClientFilter.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2017-2022 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.tracing.opentelemetry.http;
+
+import io.micronaut.context.annotation.Requires;
+import io.micronaut.core.annotation.NonNull;
+import io.micronaut.core.annotation.Nullable;
+import io.micronaut.http.HttpResponse;
+import io.micronaut.http.MutableHttpHeaders;
+import io.micronaut.http.MutableHttpRequest;
+import io.micronaut.http.annotation.Filter;
+import io.micronaut.http.client.exceptions.HttpClientResponseException;
+import io.micronaut.http.filter.ClientFilterChain;
+import io.micronaut.http.filter.HttpClientFilter;
+import io.micronaut.tracing.opentelemetry.util.TracingObserver;
+import io.micronaut.tracing.opentelemetry.util.TracingPublisherUtils;
+import io.opentelemetry.api.OpenTelemetry;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanBuilder;
+import io.opentelemetry.api.trace.Tracer;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.propagation.TextMapSetter;
+import jakarta.inject.Inject;
+import org.reactivestreams.Publisher;
+
+import static io.micronaut.tracing.opentelemetry.http.AbstractOpenTracingFilter.CLIENT_PATH;
+import static io.micronaut.tracing.opentelemetry.http.TraceRequestAttributes.CURRENT_SPAN;
+import static io.micronaut.tracing.opentelemetry.http.TraceRequestAttributes.CURRENT_SPAN_CONTEXT;
+
+/**
+ * An HTTP client instrumentation filter that uses Open Telemetry.
+ *
+ * @author Nemanja Mikic
+ * @since 1.0
+ */
+@Filter(CLIENT_PATH)
+@Requires(beans = Tracer.class)
+public class OpenTelemetryClientFilter extends AbstractOpenTracingFilter implements HttpClientFilter {
```

Comment:
It's better to use the official `opentelemetry-instrumentation-api` library here, which provides API to simplify creating new instrumentation. You can check my example which uses this library: https://github.com/tailrocks/micronaut-opentelemetry/blob/7f1e83e70e023375b3fe6684a8c273fc649d1682/opentelemetry/src/main/java/io/micronaut/opentelemetry/instrumentation/http/client/OpenTelemetryClientFilter.java#L41

If you need some explanation why it's much better than re-implement everything from scratch, please tag me and I will provide some examples.